### PR TITLE
Feature/socket keyboard

### PR DESCRIPTION
**Pull request by georgik**
Add support for keyboard via tcp socket

### CODE REVIEW

**Comment by georgik** [comment]
The implementation requires 3 threads:
- emulator thread
- server socket thread for accepting new connection
- client handler thread to process data from client

How it works:
- emulator connects to WiFi configured in `cfg.toml` and opens port 80
- client can connect from computer by Netcat: `nc IP-addr 80`
- client can send keystokes like `10p"Hello"ENTER` which is translated to `10 PRINT "Hello"ENTER`
- keystrokes from client handler thread is passed via ` std::sync::mpsc::{channel, Sender, Receiver};` to emulator thread, this seems to be very reliable even when consumer is very slow like in this case
- note: Netcat sends data only on Enter key, if you need to send data earlier just press CTRL+D to flush the buffer

This is initial sketch and it's in working state. For merge it would be beneficial to mark the network dependent part as separate feature which can user enabled during the build.

**Comment by georgik** [comment]
Few notes from refactoring
- embedded-svc must be fixed to older version, because of https://github.com/esp-rs/esp-idf-hal/issues/79
- when reference to WiFi goes out of scope, the WiFi is automatically disconnected
- tcp listener should be created inside new thread instead of passing from main thread to avoid problems with ownership

**Comment by georgik** [comment]
@SergioGasquez @mabezdev @bjoernQ @jessebraham Please review the code.
This change introduces tcpstream_keboard which allows to pass keystrokes via open TCP Socket using Netcat.
There are some challenges discovered during refactoring mentioned above.

File: src/main.rs [hunk 1]
```diff
@@ -33,44 +55,250 @@ fn main() -> Result<()> {
...
+    sys_loop_stack: Arc<EspSysLoopStack>,
+    default_nvs: Arc<EspDefaultNvs>,
+) -> anyhow::Result<Box<EspWifi>> {
+    let app_config = CONFIG;
+    let mut wifi = Box::new(EspWifi::new(netif_stack, sys_loop_stack, default_nvs)?);
+
+    wifi.set_configuration(&Configuration::Client(ClientConfiguration {
+        ssid: app_config.wifi_ssid.into(),
+        password: app_config.wifi_psk.into(),
+        auth_method: AuthMethod::None,
```
Comment (SergioGasquez) [inline src/main.rs:88]: Does using `AuthMethod::None` work on WPA2 networks? Maybe we could do some checks [like in Ferrous System's repo](https://github.com/ferrous-systems/espressif-trainings/pull/54/files)

**Comment by bjoernQ** [comment]
> when reference to WiFi goes out of scope, the WiFi is automatically disconnected

I think it's a sane decision in `esp-idf-svc` to shut down WiFi when it gets dropped

**Pull request merged**